packages/flutter/lib/src/foundation/persistent_hash_map.dart

```diff
@@ -0,0 +1,388 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+/// A collection of key/value pairs, from which you can retrieve a value
+/// using its associated key.
+///
+/// This class implements a persistent map: extending this map with a new
+/// key/value pair does not modify an existing instance but instead creates a
+/// new instance.
+///
+/// Note: unlike [Map] this class does not support `null` as a key value and
+/// implements only a functionality needed for a specific use case at the
+/// core of the framework.
+///
+/// Underlying implementation uses a variation of *hash array mapped trie*
+/// data structure with compressed (bitmap indexed) nodes.
+///
+/// See:
+///  * [Bagwell, Phil. Ideal hash trees.](https://infoscience.epfl.ch/record/64398);
+///  * [Steindorfer, Michael J., and Jurgen J. Vinju. "Optimizing hash-array mapped tries for fast and lean immutable JVM collections."](https://dl.acm.org/doi/abs/10.1145/2814270.2814312);
+///  * [Clojure's `PersistentHashMap`](https://github.com/clojure/clojure/blob/master/src/jvm/clojure/lang/PersistentHashMap.java).
+class PersistentHashMap<K extends Object, V> {
+  /// An an empty dictionary.
+  const PersistentHashMap.empty() : this._(null);
+
+  const PersistentHashMap._(this._root);
+
+  final _TrieNode? _root;
+
+  /// Create a copy of this map extended with the given [key] to [value]
+  /// mapping.
+  PersistentHashMap<K, V> copyWith(K key, V value) {
+    final _TrieNode newRoot =
+        (_root ?? _CompressedNode.empty).copyWith(0, key, key.hashCode, value);
+    if (newRoot == _root) {
+      return this;
+    }
+    return PersistentHashMap<K, V>._(newRoot);
+  }
+
+  /// Returns value associated with the given [key] or `null` if [key]
+  /// is not in the map.
+  V? lookup(K key) {
+    if (_root == null) {
+      return null;
+    }
+
+    // This is an implicit downcast to allow dart2js to omit it.
+    // ignore: return_of_invalid_type
```

Comment:
This here and elsewhere feels like a code smell...

Reply:
Tried to make it a bit less smelly by using `dart2js:as:trust` pragma instead.